pkg/cover/report.go

```diff
@@ -159,13 +159,19 @@ func (rg *ReportGenerator) generate(w io.Writer, progs []Prog, files map[string]
 		Root: new(templateDir),
 	}
 	for fname, file := range files {
-		if !strings.HasPrefix(fname, rg.buildDir) {
+		var remain string
+		if !filepath.IsAbs(fname) {
+			// If it's a relative path, it is relative to the build directory.
+			fname = filepath.Join(rg.buildDir, fname)
```

Comment:
This should be srcDir.
buildDir is where the sources were when the kernel  was built, now we could be on a different machine where source are in different dir (srcDir).
In the current impl we strip buildDir and append srcDir. Since here we have relative path, buildDir is effectively stripped already. So now we just need to append srcDir.

Reply:
I see. I thought `buildDir` was the path where the kernel was being built. In the case of fuchsia that would be `out/x64.zircon`. In our case, addr2line outputs a path relative to that folder (the fuchsia build directory). 

```
$ addr2line -afi -e ~/fuchsia/out/x64.zircon/kernel-x64-sancov/obj/kernel/zircon.elf        
ffffffff801c186e                                                                                                                                    
0xffffffff801c186e                                                                                                                                  
PcieRoot                                                                                                                                            
../../out/x64.zircon/../../zircon/kernel/dev/pcie/pcie_root.cc:13  
```

We have an open bug for this issue, but according to our toolchain team it's not easy to fix.

The ReportGenerator has no knowledge of a fuchsia build. Do you think adding an extra variable there and have this special cased for fuchsia would be fine? I would prefer keeping it as generic as possible.

Reply:
> I thought buildDir was the path where the kernel was being built. 

This is correct.
buildDir is a dir where kernel source _were_ during the build.
sourceDir is a dir where kernel sources are _now_.

If kernel was not moved, then buildDir == sourceDir.
Otherwise, buildDir != sourceDir and it is not useful to prepend buildDir (sources were moved from there), now they are in sourceDir, so we need to prepend sourceDir.

Reply:
If the source path is relative to the build directory (the one passed to pkg/vcs/fuchsia.go and pkg/build/fuchsia.go, not some other meaning of "build directory"), then we don't need to do anything special as far as I see (besides not producing an error on relative path).

Reply:
Here:

$ addr2line -afi -e ~/fuchsia/out/x64.zircon/kernel-x64-sancov/obj/kernel/zircon.elf        
ffffffff801c186e                                                                                                                                    
0xffffffff801c186e                                                                                                                                  
PcieRoot                                                                                                                                            
../../out/x64.zircon/../../zircon/kernel/dev/pcie/pcie_root.cc:13 

I presume ~/fuchsia is fuchsia checkout dir (what we will use in pkg/vcs/fuchsia).

This:

../../out/x64.zircon/../../zircon/kernel/dev/pcie/pcie_root.cc

reduces to:

../../zircon/kernel/dev/pcie/pcie_root.cc

The question is: is this path a proper source path relative to ~/fuchsia?

Reply:
> > I thought buildDir was the path where the kernel was being built.
> 
> This is correct.
> buildDir is a dir where kernel source _were_ during the build.
> sourceDir is a dir where kernel sources are _now_.

I meant "out" directory, where the kernel object files are located.


> Here:
> 
> $ addr2line -afi -e ~/fuchsia/out/x64.zircon/kernel-x64-sancov/obj/kernel/zircon.elf
> ffffffff801c186e
> 0xffffffff801c186e
> PcieRoot
> ../../out/x64.zircon/../../zircon/kernel/dev/pcie/pcie_root.cc:13
> 
> I presume ~/fuchsia is fuchsia checkout dir (what we will use in pkg/vcs/fuchsia).
> 
> This:
> 
> ../../out/x64.zircon/../../zircon/kernel/dev/pcie/pcie_root.cc
> 
> reduces to:
> 
> ../../zircon/kernel/dev/pcie/pcie_root.cc
> 
> The question is: is this path a proper source path relative to ~/fuchsia?

No. It is a source path relative to `~/fuchsia/out/x64.zircon`.

Reply:
I see. From syzkaller point of view this "out/x64.zircon" is not a build directory, it's juts a random point in file system tree (none of the build/source/object dirs will point to it). The only way to fix this is to special case fuchsia logic. I guess we either need to strip "../../" or prepend "a/b/" somewhere in pkg/cover for fuchsia then.

Reply:
ok! `pkg/cover/report` now takes the OS as a parameter, and has specific logic to handle the fuchsia case.